### ISSUE TITLE
Add Vowpal Wabbit to benchmarks

### ISSUE THREAD

**Issue by MaxHalford** [online-ml/river#147]
I haven't added Vowpal Wabbit to the benchmarks because I am incapable of making it do what I want. I tried to perform a simple logistic regression trained with plain SGD, but to no avail: 

```python
from creme import datasets
from creme import metrics
from creme import preprocessing
from vowpalwabbit import pyvw

scaler = preprocessing.StandardScaler()

metric = metrics.LogLoss()

vw = pyvw.vw(quiet=True, sgd=True, learning_rate=.005)

for x, y in datasets.fetch_electricity():
    x = scaler.fit_one(x).transform_one(x)
    y = int(y)
    ex_str = f'{y} | {" ".join(map(str, x.values()))}'
    ex = vw.example(ex_str)
    y_pred = vw.predict(ex)
    y_pred = {True: y_pred, False: 1. - y_pred}
    vw.learn(ex)  
    metric.update(y, y_pred)

metric
```

This should print output a score of 0.413533, as per [these](https://github.com/creme-ml/creme/blob/master/benchmarks/Logistic%20regression.ipynb) benchmarks. Instead if returns a score of `0.641063`. I've tried other parameters, such as `loss_function='logistic'` and `link='logistic'`, but nothing worked.

If someone could get this to work it would be really great.

**Comment by MaxHalford** [comment]
We could also add a comparison with [jax](https://github.com/google/jax).

**Comment by etiennekintzler** [comment]
Hey @MaxHalford,

I got to **0.427246** using `pyvw.vw` but unfortunately couldn't get down to 0.413533. 

Here is the code (had to change `datasets.fetch_electricity()` unto `datasets.elec2.Elec2()`, I am using `creme` version 0.6.1) : 

```python
from creme import datasets
from creme import metrics
from creme import preprocessing
from vowpalwabbit import pyvw

scaler = preprocessing.StandardScaler()

metric = metrics.LogLoss()

vw = pyvw.vw(sgd=True, learning_rate=0.005, loss_function="logistic", link="logistic")


for x, y in datasets.elec2.Elec2():
    x = scaler.fit_one(x).transform_one(x)
    y = int(y)
    y_vw = 2 * y - 1
    ex_str = f'{y_vw} | {" ".join(map(lambda x: ":" + str(x), x.values()))}'
    ex = vw.example(ex_str)
    y_pred = vw.predict(ex)
    y_pred = {True: y_pred, False: 1. - y_pred}
    vw.learn(ex)  
    metric.update(y, y_pred)

metric
```

The **differences** with your code is the following:
- You need to use the options `loss_function='logistic'` (for logistic regression) and `link='logistic'` (to output probabilities in [0, 1]).
- When using these options the labels must be {-1, 1} and not {0, 1}. If you don't do it you'll see on the console the message "You are using label 0 not -1 or 1 as loss function expects!".
- For _numerical_ feature you need to prepend it with `{name}:` or just `:` otherwise it will be considered as categorical. For example in your case  `vw.get_weight(vw.hash_feature("0.0", 0))`  will return non-null float meaning that VW learnt some weight for the categorical "0.0" (which is not the desired behavior).

By the way congrats :tada: for your library, it's quite fresh and easy to use ! Keep up the good work :+1:

**Comment by MaxHalford** [comment]
Hello @etiennekintzler! Thank you so much for those much-sought clarifications. It's nice to see that the performances are the same. I'll 1) add VW to the benchmarks 2) consider implementing that learning rate schedule. Thanks for the kind words, they always keep me motivated.

**Comment by MaxHalford** [comment]
Ok I added VW to the benchmarks. I didn't rerun the benchmarks because I'm having some issues with installing VW, but they should work. We can close this for now.

**Comment by etiennekintzler** [comment]
Did you try to pip install it or you used another way ? What error did you get ? 

Btw I just check the [Linear regression benchmark](https://github.com/creme-ml/creme/blob/master/benchmarks/Linear%20regression.ipynb) and the MSE is 3 times higher for VW, which seems quite unlikely. If you want, I can investigate to see if I get the same thing.

**Comment by MaxHalford** [comment]
I actually installed via conda, and got a segmentation fault.

If you have time to check it out that would be wonderful :). By the way, can you check if setting the `adaptive`, `normalized`, `invariant` parameters to `False` with logistic regression produces the exact score of `0.413904`?

I'll actually leave this issue open until we're 100% happy with the benchmarks.

**Comment by etiennekintzler** [comment]
> By the way, can you check if setting the adaptive, normalized, invariant parameters to False with logistic regression produces the exact score of 0.413904?

Yes they do produce the same exact score, as expected since these parameters are not used for SGD (see https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Command-line-arguments#update-rule-options): 

**--sgd**                        use regular/classic/simple stochastic  descent update, i.e., non adaptive,  normalized, non invariant                              (this is no longer the default since it is  sub-optimal)

So other parameters might be at play but I don't know which one.

----

Regarding the linear regression notebook **I got a MSE of 22.8221 with VW** (and 23.0371 with creme).

Below is a reproducible example:

```python
from creme import (stats, 
                   compose,
                   feature_extraction,
                   datasets,
                   metrics,
                   preprocessing,
                   linear_model,
                   optim)

from vowpalwabbit import pyvw

# Defining pipeline
def add_hour(x):
    x['hour'] = x['moment'].hour
    return x

pipe_prep = compose.Select('clouds', 'humidity', 'pressure', 'temperature', 'wind') + \
        (
            add_hour | \
            feature_extraction.TargetAgg(by=['station', 'hour'], how=stats.Mean())
        )\
| preprocessing.StandardScaler() 

# Defining models
lr = 0.005
linmod = linear_model.LinearRegression(
            optimizer=optim.SGD(lr),
            l2=0.,
            intercept_lr=lr
        )
vw = pyvw.vw(sgd=True, learning_rate=lr, power_t=0)

# Defining metrics
metric_vw = metrics.MSE()
metric_creme = metrics.MSE()

# Streaming through examples
for x, y in datasets.bikes.Bikes():
    
    # Preprocessing
    pipe_prep.fit_one(x=x, y=y)
    x = pipe_prep.transform_one(x)    
    
    # VW
    ex_str = f'{y} | {" ".join(map(lambda x: ":" + str(x), x.values()))}'
    ex = vw.example(ex_str)
    y_vw = vw.predict(ex)
    vw.learn(ex)  
    
    # Creme
    y_creme = linmod.predict_one(x)
    linmod.fit_one(x=x, y=y)
    
    # Computing metrics
    metric_vw.update(y, y_vw)
    metric_creme.update(y, y_creme)
    
print("Creme: ", metric_creme, 
      "\nVW: ", metric_vw)

```

**Comment by MaxHalford** [comment]
Thanks, got it. Good work on the comparison. I do wonder where the discrepancy comes from, but it's so slight that it might just be numeric instability.

Anyway now we have all the steps required to add Vowpal Wabbit to the benchmarks.

**Comment by etiennekintzler** [comment]
> Thanks, got it. Good work on the comparison. I do wonder where the discrepancy comes from, but it's so slight that it might just be numeric instability.

Yep I do wonder too haha. I check extensively the command line and did not find what could cause the variation. Also the weight initialization is 0 by default so it's not the cause either (if random it could have been).

> Anyway now we have all the steps required to add Vowpal Wabbit to the benchmarks.

Ok great !

**Comment by MaxHalford** [comment]
So I tried to install Vowpal Wabbit again today and I got some `cmake` issues that got on my nerves. I'll set up a cloud machine and try it out.

**Comment by etiennekintzler** [comment]
You might want to raise issue on vowpal wabbit project. 

Otherwise the cloud machine is a good idea. FYI it works well on my cheap (5$) Digital Ocean VM ; CPU is Intel Xeon E5-2650L v3 @ 1.80GHz, OS is Ubuntu 18.04, and python is version 3.6.9. I installed following instructions from http://vowpalwabbit.org. I can lend you the ressource if necessary.
